File: src/lnaddress/gateway.rs
use serde_json::Value;
use sha2::{Sha256, Digest};
use url::Url;
use crate::common::NwcTransport;
use crate::common::NwcError;
use crate::lnaddress::wallet_connector::NwcSession;

pub fn pay_request_info(username: &str, request_url: &Url) -> Value {
    let callback_url = build_callback_url(username, request_url);

    serde_json::json!({
        "callback": callback_url,
        "maxSendable": 100000000,
        "minSendable": 1000,
        "metadata": generate_metadata(username),
        "tag": "payRequest"
    })
}

pub async fn create_invoice(
    transport: &impl NwcTransport,
    nwc_uri: &str,
    username: &str,
    amount_msat: u64,
) -> Result<String, NwcError> {
    let description_hash = get_description_hash(username);
    let session = NwcSession::new(transport, nwc_uri)?;
    session.make_invoice(amount_msat, description_hash).await
}

fn generate_metadata(username: &str) -> String {
    format!("[[\"text/plain\",\"Payment to {}\"]]", username)
}

fn get_description_hash(username: &str) -> String {
    let metadata = generate_metadata(username);
    let mut hasher = Sha256::new();
    hasher.update(metadata.as_bytes());
    let hash = hasher.finalize();
    hex::encode(hash)
}

fn build_callback_url(username: &str, request_url: &Url) -> String {
    let is_local = request_url.host_str() == Some("localhost")
        || request_url.host_str() == Some("127.0.0.1");
    let host = request_url.host_str().unwrap_or_default();
    let port = request_url
        .port()
        .map(|p| format!(":{}", p))
        .unwrap_or_default();
    let protocol = if is_local { "http" } else { "https" };

    format!(
        "{}://{}{}/lnaddress/{}/callback",
        protocol, host, port, username
    )
}

#[cfg(test)]
#[path = "gateway_test.rs"]
mod gateway_test;

File: src/lnaddress/gateway.ts
import type { NwcTransport } from '../common/mod.ts';
import { sha256 } from '@noble/hashes/sha2.js';
import { hexEncode } from '../util.ts';
import { NwcSession } from './wallet_connector.ts';

export function payRequestInfo(username: string, requestUrl: URL): Record<string, unknown> {
  const callbackUrl = buildCallbackUrl(username, requestUrl);

  return {
    callback: callbackUrl,
    maxSendable: 100_000_000,
    minSendable: 1000,
    metadata: generateMetadata(username),
    tag: 'payRequest',
  };
}

export async function createInvoice(
  transport: NwcTransport,
  nwcUri: string,
  username: string,
  amountMsat: number,
): Promise<string> {
  const descriptionHash = getDescriptionHash(username);
  const session = new NwcSession(transport, nwcUri);
  return session.makeInvoice(amountMsat, descriptionHash);
}

export function generateMetadata(username: string): string {
  return `[["text/plain","Payment to ${username}"]]`;
}

export function getDescriptionHash(username: string): string {
  const metadata = generateMetadata(username);
  const hash = sha256(new TextEncoder().encode(metadata));
  return hexEncode(hash);
}

export function buildCallbackUrl(username: string, requestUrl: URL): string {
  const isLocal = requestUrl.hostname === 'localhost' || requestUrl.hostname === '127.0.0.1';
  const host = requestUrl.hostname;
  const port = requestUrl.port ? `:${requestUrl.port}` : '';
  const protocol = isLocal ? 'http' : 'https';

  return `${protocol}://${host}${port}/lnaddress/${username}/callback`;
}
